### PR TITLE
More trigger fixes

### PR DESCRIPTION
**Pull request by brycemcanally**
This PR fixes a few issues related to triggers that were identified by issues with TestUpgradeTrigger. The first fix is to make triggers idempotent. Triggers are evaluated / fired during commit finishing and commit finishing must be idempotent since it isn't guaranteed to only occur once for a particular commit (PFS masters acquire a distributed lock that doesn't provide strict mutual exclusion). Since triggers were not idempotent, they could be fired more than once for a particular commit. The second fix was to make the commit count trigger ignore the initial empty commit that is created when a branch is created. This just isn't behavior that a user would expect. The third fix was to update the TestUpgradeTrigger test to do an additional commit to account for the off-by-one error in the old versions, wait for the triggers to fire for the existing commits before upgrading, and count the commits created by the triggers to ensure that the correct number of commits are created (before the idempotent change, there would be more commits than expected here). There is still some investigation needed for why commit finishing is firing multiple times because that should be a rare edge case when the distributed lock is lost.

### CODE REVIEW

**Comment by FahadBSyed** [comment]
These changes look good, however it looks like there is another test that is breaks `TestPFS/Trigger/Count`:
```
Failed
=== RUN   TestPFS/Trigger/Count
=== PAUSE TestPFS/Trigger/Count
=== CONT  TestPFS/Trigger/Count
    server_test.go:6752: Not equal: "116c041429c54caca49c59ea8b2f8f4e" (expected)
                != "4363fc1d43c846518e3f5fba9c4712da" (actual)
    server_test.go:6752: current stack:
        goroutine 1382 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x5e
        github.com/pachyderm/pachyderm/v2/src/internal/require.fatal({0x403bf48, 0xc001bbcd00}, {0x0, 0x0, 0x0}, {0xc0006823f0, 0x6f}, {0x0, 0x0, 0x0})
        	/home/circleci/project/src/internal/require/require.go:599 +0x19a
        github.com/pachyderm/pachyderm/v2/src/internal/require.Equal({0x403bf48, 0xc001bbcd00}, {0x3025860, 0xc00178f520}, {0x3025860, 0xc00178f530}, {0x0, 0x0, 0x0})
        	/home/circleci/project/src/internal/require/require.go:79 +0x112
        github.com/pachyderm/pachyderm/v2/src/server/pfs/server/testing.TestPFS.func147.5(0xc001bbcd00?)
        	/home/circleci/project/src/server/pfs/server/testing/server_test.go:6752 +0xdb7
        testing.tRunner(0xc001bbcd00, 0xc00190fba0)
        	/usr/local/go/src/testing/testing.go:1595 +0xff
        created by testing.(*T).Run in goroutine 123
        	/usr/local/go/src/testing/testing.go:1648 +0x3ad
--- FAIL: TestPFS/Trigger/Count (0.31s)
```

File: src/server/pfs/server/trigger.go [hunk 1]
```diff
@@ -131,16 +131,15 @@ func (d *driver) isTriggered(txnCtx *txncontext.TransactionContext, t *pfs.Trigg
...
-			} else {
+			if oldHead.Commit.Id == ci.Commit.Id || ci.ParentCommit == nil {
 				break
 			}
+			var err error
+			ci, err = d.resolveCommit(txnCtx.SqlTx, ci.ParentCommit)
+			if err != nil {
+				return false, err
+			}
+			commits++
```
Comment (FahadBSyed) [inline src/server/pfs/server/trigger.go:142]: By moving the count to the end, the code will break before it counts the initial commit (because the initial commit's id == ci.commit.ID)

File: src/testing/deploy/upgrade_test.go [hunk 2]
```diff
@@ -112,36 +112,55 @@ func TestUpgradeTrigger(t *testing.T) {
...
+			})
 		},
 		func(t *testing.T, ctx context.Context, c *client.APIClient, _ string) { /* postUpgrade */
 			for i := 0; i < 10; i++ {
 				require.NoError(t, c.PutFile(dataCommit, "/hello", strings.NewReader("hello world")))
 			}
 			latestDataCI, err := c.InspectCommit(pfs.DefaultProjectName, dataRepo, "master", "")
 			require.NoError(t, err)
-			require.NoErrorWithinTRetry(t, 2*time.Minute, func() error {
+			require.NoErrorWithinTRetry(t, 5*time.Minute, func() error {
```
Comment (FahadBSyed) [inline src/testing/deploy/upgrade_test.go:146]: is the timer increase still needed?

**Review by FahadBSyed — CHANGES_REQUESTED** [review]
These changes look good, however it looks like there is another test that is breaks `TestPFS/Trigger/Count`:
```
Failed
=== RUN   TestPFS/Trigger/Count
=== PAUSE TestPFS/Trigger/Count
=== CONT  TestPFS/Trigger/Count
    server_test.go:6752: Not equal: "116c041429c54caca49c59ea8b2f8f4e" (expected)
                != "4363fc1d43c846518e3f5fba9c4712da" (actual)
    server_test.go:6752: current stack:
        goroutine 1382 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x5e
        github.com/pachyderm/pachyderm/v2/src/internal/require.fatal({0x403bf48, 0xc001bbcd00}, {0x0, 0x0, 0x0}, {0xc0006823f0, 0x6f}, {0x0, 0x0, 0x0})
        	/home/circleci/project/src/internal/require/require.go:599 +0x19a
        github.com/pachyderm/pachyderm/v2/src/internal/require.Equal({0x403bf48, 0xc001bbcd00}, {0x3025860, 0xc00178f520}, {0x3025860, 0xc00178f530}, {0x0, 0x0, 0x0})
        	/home/circleci/project/src/internal/require/require.go:79 +0x112
        github.com/pachyderm/pachyderm/v2/src/server/pfs/server/testing.TestPFS.func147.5(0xc001bbcd00?)
        	/home/circleci/project/src/server/pfs/server/testing/server_test.go:6752 +0xdb7
        testing.tRunner(0xc001bbcd00, 0xc00190fba0)
        	/usr/local/go/src/testing/testing.go:1595 +0xff
        created by testing.(*T).Run in goroutine 123
        	/usr/local/go/src/testing/testing.go:1648 +0x3ad
--- FAIL: TestPFS/Trigger/Count (0.31s)
```

We can merge these changes in now and open up a new PR for the other test, or we can include the fix for the second test in these changes too. For now I am marking as needs work until we explicitly reach a decision.

File: src/server/pfs/server/trigger.go [hunk 3]
```diff
@@ -131,16 +131,15 @@ func (d *driver) isTriggered(txnCtx *txncontext.TransactionContext, t *pfs.Trigg
 		ci := newHead
 		var commits int64
 		for commits < t.Commits {
```
Comment (albscui) [inline src/server/pfs/server/trigger.go:133]: I think it's a bit cleaner to write this loop like

```Go
for ci.ParentCommit != nil && commits < t.Commits {
			if ci.Commit.Id == oldHead.Commit.Id {
				break
			}
			var err error
			ci, err = d.resolveCommit(ctx, txnCtx.SqlTx, ci.ParentCommit)
			if err != nil {
				return false, err
			}
			commits++
		}
```

File: src/server/pfs/server/trigger.go (hunk 1 above)
Comment (albscui) [reply src/server/pfs/server/trigger.go:142]: That's not a problem. The loop invariant is that `newHead != oldHead`. We terminate the loop early if that is no longer true, so we don't want to increment the count before that.

File: src/server/pfs/server/trigger.go (hunk 3 above)
Comment (brycemcanally) [reply src/server/pfs/server/trigger.go:133]: To me that seems less clear, but that may just be how I read it. I read it as traverse t.Commits, but break out if the oldHead or empty commit we create is reached. What you proposed just pulls one of the break conditions up to the loop conditional. It seems like you might as well put the full conditional there if you do that.

**Review by albscui — APPROVED** [review]
